errorcodes/errorcodes.go

```diff
@@ -0,0 +1,241 @@
+/*
+Copyright 2023 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package errorcodes
+
+import (
+	"errors"
+
+	"google.golang.org/genproto/googleapis/rpc/errdetails"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/runtime/protoiface"
+)
+
+const (
+	ErrorCodesFeatureMetadataKey = "error_codes_feature"
+	Owner                        = "components-contrib"
+	Domain                       = "dapr.io"
+	NoReason                     = "NO_REASON_SPECIFIED"
+)
+
+var NoErrorReason = WithErrorReason(NoReason, 503, codes.Internal)
+
+type ResourceInfo struct {
+	ResourceType string
+	ResourceName string
+}
+
+func WithErrorReason(reason string, httpCode int, grpcStatusCode codes.Code) ErrorOption {
```

Comment:
Can't this be a function of DaprError instead?

Reply:
The idea of the `WithErrorReason` is to provide the ability to create a `ErrorOption` values which are used as part of the DaprError construction. This is how I envision it'll be used within components-contrib
```golang
var (
	StateETagMismatchReason = errorcodes.WithErrorReason(StateStoreEtagMismatch, 400, codes.Aborted)
	StateETagInvalidReason  = errorcodes.WithErrorReason(StateStoreETagInvalid, 400, codes.Aborted)
)
...
if de := state.NewDaprError(err, req.Metadata, state.StateETagMismatchReason); de != nil {
				de.SetResourceInfo(resourceInfo)
				de.SetDescription(fmt.Sprintf("state store Set - possible etag(%s) %s. original error: %v", *req.ETag, string(state.ETagMismatch), err))
				return de
}
```